### PR TITLE
fix: Add explicit null-buffer guards and shared-memory fallback for Intel Macs with discrete GPU (MPS)

### PR DESCRIPTION
**Pull request by bobbyshop-vui**
# Changelog: MPS Intel/Non-Apple-Silicon Compatibility Patch

Full list of files changed in this branch, based on the actual diff. This
replaces the earlier README, which only covered the 2 files I authored
(`MPSAllocator.h` / `MPSAllocator.mm`) — this version covers all 7.

---

## 1. `aten/src/ATen/mps/MPSAllocator.h`

**+159 / -17**

- `UsageFlags::SHARED` / `UsageFlags::MANAGED`: comments rewritten to state
  that `SHARED` is now only ever used on genuine Apple Silicon, and
  `MANAGED` is the Intel/AMD substitute — enforced via the new
  `normalizeUsage()`.
- `kHostAccessibleUsageMask`: comment updated (`SHARED | MANAGED`,
  unchanged value).
- `BufferBlock::cpu_ptr` comment: now mentions Shared **or** Managed.
- `HeapBlock::createHeapBlock()`:
  - `d.storageMode` ternary extended to check `MANAGED` first, then
    `SHARED`, then falls back to `MTLStorageModePrivate`.
  - The line `d.resourceOptions = getOptions(usage);` **removed**, with a
    comment explaining it's a legacy/deprecated property that conflicts
    with the explicit `storageMode`/`cpuCacheMode`/`hazardTrackingMode`
    already set, and was the cause of an `Invalid storageMode (Invalid)`
    heap validation failure.
- `BufferPool::usage` comment: clarified that `SHARED_SMALL`/`SHARED_LARGE`
  are pool *role* names, not a literal storage-mode guarantee — on
  Intel/AMD they're actually Managed.
- `MPSHeapAllocatorImpl` constructor:
  - `m_has_unified_memory` now requires **both**
    `[m_device hasUnifiedMemory]` **and**
    `[m_device supportsFamily:MTLGPUFamilyApple1]`, with a comment
    explaining that Intel integrated GPUs also report `hasUnifiedMemory ==
    YES` and would otherwise be misclassified.
  - `m_max_buffer_size` now guards against `m_device == nil`
    (`m_device != nil ? [m_device maxBufferLength] : 0`).
- Destructor: now also releases `m_scalar_staging_buffer` if non-nil.
- `isSharedBuffer()` doc comment: "Shared on Apple Silicon, Managed on
  Intel/AMD" (was: "shared pool" only).
- `getSharedBufferPtr()` / `getHostAliasStorage()` / `recordEvents()` /
  `waitForEvents()` doc comments: "shared" → "host-accessible
  (Shared/Managed)" throughout.
- `hasUnifiedMemory()` accessor comment: clarified semantics (true only on
  Apple-family + unified memory, independent of the raw API flag).
- `default_high_watermark_ratio` / `default_low_watermark_ratio` comments:
  labeled as the Apple Silicon defaults; new `_intel` counterparts
  (`default_high_watermark_ratio_intel = 1.0`,
  `default_low_watermark_ratio_intel = 0.95`) added with comments
  explaining the more conservative ratios for a small, fixed discrete-GPU
  VRAM pool.
- `m_has_unified_memory` member: comment updated to reference the
  corrected constructor logic.
- New private members/methods:
  - `id<MTLBuffer> m_scalar_staging_buffer = nil;`
  - `hostAccessibleUsage()`: now returns `PRIVATE` (not `MANAGED`) on
    non-Apple-Silicon, with a comment explaining that `MTLHeap` disallows
    `MTLStorageModeManaged` entirely (per Apple's own `MTLHeap.h`), so
    Managed was never a valid *heap* fallback in the first place.
  - `normalizeUsage(uint32_t usage)`: funnel that remaps any
    SHARED/MANAGED request to whatever `hostAccessibleUsage()` says the
    device's heaps can back.
  - `uploadScalarViaBlit(id<MTLBuffer> dst, const void* value, size_t
    size)`: declared — uploads a CPU value into a Private heap buffer via
    a standalone Managed staging buffer + blit, for non-Apple-Silicon
    devices.
  - `synchronizeManagedBuffer()` comment: updated to note it's now mostly
    relevant only to the standalone staging buffer, since no heap-backed
    buffer is ever Managed any more.
- `max_device_size()` / `current_allocated_size()`: both now guard against
  `m_device == nil`, returning `0` instead of messaging a nil device.

---

## 2. `aten/src/ATen/mps/MPSAllocator.mm`

**+172 / -31** (large diff, not rendered inline by GitHub — based on what
was described/applied directly with you in this conversation)

- `init_allocator()`: comment updated to describe the Private fallback (not
  Managed) for non-Apple-Silicon heaps.
- `init_buffer_pools()`: comment rewritten around the corrected
  Private/Shared-only heap model; pools still constructed via
  `hostAccessibleUsage()`.
- `alloc_buffer_block()`: now calls `usage = normalizeUsage(usage);` before
  computing allocation size / picking a pool — the single funnel point
  that guarantees Intel/AMD never requests a Shared- or (heap-backed)
  Managed-storage heap, regardless of which literal flag the caller passed.
- `allocScalarBufferWithValue()`: reworked to branch on
  `m_has_unified_memory`:
  - Apple Silicon: unchanged direct `memcpy` into `cpu_ptr`.
  - Everywhere else: delegates to the new `uploadScalarViaBlit()` instead
    of writing to `cpu_ptr` directly (the scalar pool is now Private there,
    so it has no CPU pointer at all).
- New `uploadScalarViaBlit()` implementation: lazily creates
  `m_scalar_staging_buffer` (`MTLResourceStorageModeManaged`, sized
  `kMaxScalarAlloc`), `memcpy`s the value in, calls `didModifyRange:`, then
  blits from staging into the destination buffer via
  `MTLBlitCommandEncoder`, followed by `synchronize(COMMIT_AND_WAIT)`.
- `isSharedBuffer()` / `getSharedBufferPtr()` / `getHostAliasStorage()`:
  comments updated ("Shared on Apple Silicon, Managed elsewhere" /
  "non-Apple-Silicon (Managed storage)").
- `synchronizeManagedBuffer()`: comment updated; logic unchanged (no-op on
  `m_has_unified_memory`, blit-and-wait otherwise).
- `getIMPSAllocator()` / `isMPSPinnedPtr()` doc comments: "Intel" →
  "Intel/AMD" / "every other device" for accuracy.

---

## 3. `aten/src/ATen/mps/MPSDevice.h`

**+1 / -1**

- `MacOSVersion::MACOS_VER_14_4_PLUS = 0` → `MACOS_VER_14_4_PLUS` (drops
  the explicit `= 0`). Purely cosmetic — first enumerator defaults to `0`
  either way; no behavior change.

---

## 4. `aten/src/ATen/mps/MPSDevice.mm`

**+1 / -4**

- MPS-availability probe in the `MPSDevice` constructor changed from:
  ```objc
  if ([mpsCD instancesRespondToSelector:@selector(HermiteanToRealFFTWithTensor:axes:descriptor:name:)] == NO)
  ```
  to:
  ```objc
  if (mpsCD == nil)
  ```
  i.e. from "does the `MPSGraph` class respond to a specific (fairly
  recent) selector" to "does the `MPSGraph` class exist at all". This is a
  **much weaker** capability check — it will treat older `MPSGraph`
  versions that lack that selector (and the operations built on it) as
  fully capable, where the old code would have correctly excluded them.
- The device-enumeration loop:
  ```objc
  if ([device isLowPower]) { // exclude Intel GPUs
    continue;
  }
  ```
  was **removed**. This is the line that used to skip Intel GPUs
  entirely — removing it is what allows `MPSDevice` to pick an Intel GPU
  as `_mtl_device` at all, which is a prerequisite for any of this Intel
  support to work. Intentional and consistent with the goal of this
  branch, but worth flagging explicitly since it changes device selection
  for *every* Mac with both an Intel iGPU and a discrete GPU, not just
  Intel-only Macs.

---

## 5. `aten/src/ATen/mps/MPSHooks.mm`

**+7 / -1**

```objc
bool MPSHooks::hasMPS() const {
  return at::mps::is_available();
}
```
changed to:
```objc
bool MPSHooks::hasMPS() const {
  static bool available = []() {
    @autoreleasepool {
      NSOperatingSystemVersion version = [[NSProcessInfo processInfo] operatingSystemVersion];
      return version.majorVersion > 10 || (version.majorVersion == 10 && version.minorVersion >= 15);
    }
  }();
  return available;
}
```

⚠️ **Flagged as a likely regression, not yet resolved.** This is the
public entry point for `torch.backends.mps.is_available()`. The new
version checks **only the macOS version** (≥ 10.15) and no longer calls
`at::mps::is_available()`, which actually checks whether a real
`MTLDevice` was found (`MPSDevice::getInstance()->device() != nil`).

Net effect: `hasMPS()` can now return `true` on a Mac running macOS 10.15+
even if no usable Metal device was found at all (no GPU, headless VM,
etc.) — exactly the nil-device situation the allocator's new
`m_device != nil` guards in `MPSAllocator.h`/`.mm` were added to protect
against. This looks inconsistent with the rest of the patch and should
probably call `at::mps::is_available()` (or something that still checks
for a real device) in addition to / instead of the OS-version check.

---

## 6. `aten/src/ATen/native/mps/operations/Activation.mm`

**+2 / -2**

```objc
TORCH_IMPL_FUNC(gelu_out_mps)(const Tensor& self, std::string_view approximate, const Tensor& output) {
  GeluKernel(kMPS, *this, get_gelutype_enum(approximate));
}
TORCH_IMPL_FUNC(gelu_backward_out_mps)(...) {
  GeluBackwardKernel(kMPS, *this, get_gelutype_enum(approximate));
}
```
changed to:
```objc
TORCH_IMPL_FUNC(gelu_out_mps)(const Tensor& self, std::string_view approximate, const Tensor& output) {
}
TORCH_IMPL_FUNC(gelu_backward_out_mps)(...) {
}
```

⚠️ **Flagged as broken, not a fix.** Both function bodies are now empty —
the calls to `GeluKernel(...)` / `GeluBackwardKernel(...)` were deleted,
not replaced with anything. As written, `output` and `grad_input` are
never written to, so `torch.nn.functional.gelu` on MPS would silently
produce whatever garbage was already in the output tensor's memory,
instead of either working correctly or raising an error. This needs to be
reverted or replaced with real logic before merging — please confirm
whether this was intentional (e.g. a deliberate stub while something else
is reworked) or a leftover from a merge conflict / accidental deletion.

---

## 7. `aten/src/ATen/native/mps/OperationUtils.mm`

**+1 / -1**

```objc
} else {
  [options setLanguageVersion:MTLLanguageVersion3_1];
}
```
changed to:
```objc
} else {
  [options setLanguageVersion:MTLLanguageVersion3_0];
}
```

This is the fix for the `Function ... is using language version 3.1 which
is incompatible with this OS` crash seen when running training on this
hardware: the `else` branch (OS that doesn't qualify for the newer
`MTLLanguageVersion3_2` branch above it) was hardcoded to request MSL
**3.1**, which this Mac's OS doesn't support; it now requests MSL **3.0**
instead.

⚠️ **On its own this is necessary but not sufficient.** Compiling at
`-std=metal3.0` only works if no `.metal` source or shared header
(`c10/metal/atomic.h`, `c10/metal/utils.h`, `c10/metal/indexing.h`, etc.)
references the `bfloat` type or other MSL-3.1-only features, since
`bfloat` doesn't exist before MSL 3.1. The build errors seen earlier
(`use of undeclared identifier 'bfloat'`, `redefinition of
'vectypes<float>'`) confirm those headers still need separate handling —
this single-line change doesn't address them.

---

## Open items before this branch is safe to merge/use

1. **`Activation.mm`** — restore or properly reimplement `gelu_out_mps` /
   `gelu_backward_out_mps`; they currently do nothing.
2. **`MPSHooks.mm`** — `hasMPS()` should verify an actual Metal device was
   found, not just the macOS version.
3. **`bfloat` in `c10/metal/{atomic,utils,indexing}.h`** — still needs to
   be guarded out (e.g. `#if __METAL_VERSION__ >= 310`) or stripped for
   MSL-3.0 builds, consistent with whatever the `.metal` kernel files
   already do.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review these changes

**Comment by bobbyshop-vui** [comment]
wait that an error i will fix code now

**Review by aorenste — DISMISSED** [review]
I'm not sure if this is a bad merge or what but this changes WAY more than described in the summary.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by bobbyshop-vui** [comment]
Wait i try to compile again in my mac that get crash and i need to fix sorry

**Comment by bobbyshop-vui** [comment]
I am done

**Comment by bobbyshop-vui** [comment]
@claude review these changes
